fms/utils/tensors.py

```diff
@@ -129,9 +132,253 @@ def cat(tensors, dim=0, *, out=None):
+
+        decrement(page_ids)
+
+
+class PagedTensor(torch.Tensor):
+    """
+    A PagedTensor stores tensor data in pages of tensor memory. This allows for
+    allocating more space than required, and appending data into an existing
+    tensor. It can also allow creating more sophisticated views of the
+    underlying storage, which can be useful for kv caching with speculative
+    decoding (because we decode many possible continuations of the same cache).
+
+    Args:
+    tensor: the tensor to use in paged storage
+    static_dims: The dims that won't vary, e.g. an embedding dimension
+    page_size: the number of entries of the static portion of the tensor to
+        keep in one "page"
+    """
+
+    def __init__(
+        self,
+        tensor: torch.Tensor,
+        static_dims: Union[int, List[int]] = -1,
+        page_size: int = 16,
+    ):
+        super().__init__()
+
+        if isinstance(static_dims, int):
+            self.static_dims = [static_dims]
+        else:
+            self.static_dims = static_dims
+
+        # normalize the dims in case of dims like `-1`
+        self.static_dims = [dim % tensor.ndim for dim in self.static_dims]
+        self.dynamic_dims = [
+            i for i, _ in enumerate(tensor.shape) if i not in self.static_dims
+        ]
+        self.current_shape = list(tensor.shape)
+        self.page_size = page_size
+        self.static_shape = [tensor.shape[x] for x in self.static_dims]
+        self.dynamic_shape = [tensor.shape[x] for x in self.dynamic_dims]
+        self.true_page_size = page_size * math.prod(self.static_shape)
+        self.min_pages = tensor.numel() // self.true_page_size
+        if self.min_pages * self.true_page_size != tensor.numel():
+            self.min_pages += 1
+
+        self.page_ids, _ = _create_page_ids(self.dynamic_shape, page_size, 0)
+        self.paged_storage = PagedStorage(
+            self.static_shape,
+            self.dynamic_shape,
+            self.page_size,
+            tensor,
+            self.page_ids,
+            dtype=tensor.dtype,
+            device=tensor.dtype,
+        )
+
+    def __new__(cls, tensor, static_dims=-1, page_size=16):
+        return super().__new__(cls)
+
+    def __del__(self):
+        self.paged_storage.decrement_refcounts(self.page_ids)
+
+    def size(self, dim=None):
+        # https://github.com/pytorch/pytorch/issues/111944
+        if dim is None:
+            return self.current_shape
+        else:
+            return self.current_shape[dim]
+
+    def _tensor(self):
+        """
+        Returns a copy of the tensor, copied from paged storage
+        """
+        result = torch.empty(
+            self.current_shape,
+            device=self.paged_storage.storage.device,
+            dtype=self.paged_storage.storage.dtype,
+        )
+
+        ranges = [range(dim) for dim in self.dynamic_shape[:-1]]
+        dynamic_indices = list(itertools.product(*ranges))
+
+        for indices in dynamic_indices:
```

Comment:
Add comments to this loop creating the tensor

Reply:
added